source_integration_test.go

```diff
@@ -0,0 +1,410 @@
+	is.NoErr(err)
+
+	err = source.Open(context.Background(), nil)
+	is.NoErr(err)
+
+	// snapshot, one record
+	rec, err := source.Read(ctx)
+	is.NoErr(err)
+	is.Equal(rec.Payload.After, opencdc.StructuredData{PartitionKey: "pkey1", SortKey: "1"})
+
+	// add a row, will be captured by CDC
+	err = insertRecord(ctx, client, testTable, 2, 3)
+	is.NoErr(err)
+
+	// delete the latest row, will be captured by CDC
+	_, err = client.DeleteItem(ctx, &dynamodb.DeleteItemInput{
+		TableName: aws.String(testTable),
+		Key: map[string]types.AttributeValue{
+			PartitionKey: &types.AttributeValueMemberS{Value: "pkey2"},
+			SortKey:      &types.AttributeValueMemberN{Value: "2"},
+		},
+	})
+	is.NoErr(err)
+
+	// update the first row, will be captured by CDC
+	_, err = client.UpdateItem(ctx, &dynamodb.UpdateItemInput{
+		TableName: aws.String(testTable),
+		Key: map[string]types.AttributeValue{
+			PartitionKey: &types.AttributeValueMemberS{Value: "pkey1"}, // partition key
+			SortKey:      &types.AttributeValueMemberN{Value: "1"},     // sort key
+		},
+		UpdateExpression: aws.String("SET #attr = :newValue"),
+		ExpressionAttributeNames: map[string]string{
+			"#attr": "data", // alias 'data' to avoid reserved keyword error
+		},
+		ExpressionAttributeValues: map[string]types.AttributeValue{
+			":newValue": &types.AttributeValueMemberS{Value: "newValue"}, // the new value
+		},
+	})
+	is.NoErr(err)
+
+	// cdc
+	i := 0
+	for {
+		rec2, err := source.Read(ctx)
+		if err == nil {
+			switch i {
+			case 0:
+				is.True(rec2.Operation == opencdc.OperationCreate)
+				is.Equal(rec2.Payload.After, opencdc.StructuredData{PartitionKey: "pkey2", SortKey: "2"})
+			case 2:
+				is.True(rec2.Operation == opencdc.OperationDelete)
+				is.Equal(rec2.Payload.Before, opencdc.StructuredData{PartitionKey: "pkey2", SortKey: "2"})
+			case 1:
+				is.True(rec2.Operation == opencdc.OperationUpdate)
+				is.Equal(rec2.Payload.After, opencdc.StructuredData{"data": "newValue", PartitionKey: "pkey1", SortKey: "1"})
+			}
+		}
+		i++
+		if i == 3 {
+			break
+		}
+	}
+
+	_ = source.Teardown(ctx)
+}
+
+func prepareIntegrationTest(ctx context.Context, t *testing.T) (*dynamodb.Client, map[string]string) {
+	t.Helper()
+	cfg := map[string]string{
+		SourceConfigAwsAccessKeyId:     "test",
+		SourceConfigAwsSecretAccessKey: "test",
+		SourceConfigAwsRegion:          "us-east-1",
+		SourceConfigPollingPeriod:      "10ms",
+		SourceConfigAwsUrl:             "http://localhost:4566", // docker url
+	}
+
+	client, err := newDynamoClients(ctx, cfg)
+	if err != nil {
+		t.Fatalf("could not create dynamoDB clients: %v", err)
+	}
+
+	table := "conduit-dynamodb-source-test-" + uuid.NewString()
+	cfg[SourceConfigTable] = table
+
+	// create table
+	err = createTable(ctx, client, table, PartitionKey, SortKey)
+	if err != nil {
+		t.Fatalf("could not create dynamoDB table: %v", err)
+	}
+
+	t.Cleanup(func() {
+		clearTable(ctx, client, table, PartitionKey, SortKey)
+		deleteTable(ctx, client, table)
```

Comment:
We don't check the errors here. I propose to collect them and log the error, if any.

```suggestion
		var errs []error
		errs = append(errs, clearTable(ctx, client, table, PartitionKey, SortKey))
		errs = append(errs, deleteTable(ctx, client, table))
		if err := errors.Join(errs); err != nil {
			t.Logf("failed to clear and delete table: %s", err)
		}
```


Btw, why do we clear the table before deleting it? Is that required by DynamoDB?

Reply:
that's what ChatGPT told me.. until I just checked the docs and it's not needed 🫠